addons/frag/functions/fnc_doFragTargeted.sqf

```diff
@@ -0,0 +1,177 @@
+#include "..\script_component.hpp"
+/*
+ * Author: Jaynus, NouberNou, Lambda.Tiger
+ * This function creates fragments targeted at specific entities, up to
+ * a configured maximum
+ *
+ * Arguments:
+ * 0: Position of fragmenting projectile ASL <ARRAY>
+ * 1: Velocity of the fragmenting projectile <ARRAY>
+ * 2: Maximum range of fragments to calculate <SCALAR>
+ * 3: Maximum number of fragments to produce <SCALAR>
+ * 4: Types of fragments <ARRAY>
+ * 5: A modified parameter used to calculate whether a fragment hits <SCALAR>
+ * 6: Shot parent <ARRAY>
+ *
+ * Return Value:
+ * Number of fragments created <SCALAR>
+ *
+ * Example:
+ * [getPosASL _proj, velocity _proj, 50, 50, [], 1, [player, player]] call ace_frag_fnc_doFragTargeted;
+ *
+ * Public: No
+ */
+
+params [
+    "_posASL",
+    ["_fragVelocity", 800, [123]],
+    ["_fragRange", 50, [123]],
+    ["_maxFrags", 20, [123]],
+    ["_fragTypes", [], [[]]],
+    ["_modFragCount", 1, [123]],
+    ["_shotParents", [objNull, objNull], [[]], [2]]
+];
+TRACE_5("fnc_doFragTargeted",_posASL,_fragRange,_maxFrags,_fragTypes,_modFragCount);
+
+if (_fragTypes isEqualTo []) then {
+    _fragTypes = [
+        QGVAR(tiny), QGVAR(tiny), QGVAR(tiny),
+        QGVAR(tiny_HD), QGVAR(tiny_HD), QGVAR(tiny_HD),
+        QGVAR(small), QGVAR(small), QGVAR(small), QGVAR(small),
+        QGVAR(small_HD), QGVAR(small_HD), QGVAR(small_HD), QGVAR(small_HD),
+        QGVAR(medium_HD), QGVAR(medium_HD), QGVAR(medium_HD), QGVAR(medium_HD), QGVAR(medium_HD)
+    ];
+};
+
+private _objects = (ASLToATL _posASL) nearEntities [["Car", "Motorcycle", "Tank", "StaticWeapon", "CAManBase", "Air", "Ship"], _fragRange];
+if (_objects isEqualTo []) exitWith {
+    TRACE_2("No nearby targets",_posASL,_fragRange);
+    0;
+};
+
+// grab crews and add them in so that targets stay approx. sorted by distance
+{
+    private _crew = crew _x;
+    if (count _crew > 1) then {
+        private _arr = [_x];
+        {
+            _arr pushBackUnique _x;
+        } forEach _crew;
+
+        _objects set [_forEachIndex, _arr];
+    };
+} forEach _objects;
+_objects = flatten _objects;
```

Comment:
Are you saying to do _objects append _crew? It is faster, but with the algorithm for generating fragments, it'd prefer objects that are far over a units in go-carts next to the fragmentation event. I thought it was worth the extra time for the preference. If I misunderstood, would you mind clarifying?

Reply:
This is faster:

```
private _targets = [];

{
    private _crew = crew _x;
    
    if (_crew isNotEqualTo []) then {
        _crew pushBackUnique _x;
        _targets append _crew;
    } else {
        _targets pushBack _x;
    };
} forEach _objects;

_targets // use _targets instead of _objects

```

Reply:
`crew` returns the unit, if the unit is on foot. It's wasteful, so it might be better to get `nearEntities` for vehicles and units separately (not sure of performance impact though).

Reply:
Ah, apologies I misunderstood that. I'll do some tests and commit the best approach.

Reply:
Alright, I tested this on an area of 100m which is typical for medium fragement count events like western hand grenades (for some reason western arms manufacturers really prefers more, low-mass high-energy fragments). Area contained 58 targets including 2 crewed vehicles and 5 empty and scattered other objects:
![image](https://github.com/acemod/ACE3/assets/11843718/266a4d98-581c-4c48-86e5-62949a209d86)

This is is fastest while keeping the ordering by distance mostly right with a runtime of about 40% of the current code. Oddly enough, checking if `_crew` is empty is way slower than the method below
```sqf
private _objects = (ASLtoATL _posASL) nearEntities [["CAManBase","Car", "Motorcycle", "Tank", "StaticWeapon", "Air", "Ship"], _fragPower];    
  
private _targets = [];   
{   
    private _crew =  crew _x;
    _crew pushBackUnique _x;
    _targets append _crew;   
  
} forEach _objects;   
```
While if we life with imperfect ordering of dismounted infantry, then crews & their vehicles, execution time goes to about ~27% of the current code as long as the number of vehicles is low.
```sqf
private _posATL = ASLtoATL _posASL;
private _targets =_posATL nearEntities [["CAManBase"], _fragRange];  
private _vehicles = _posATL nearEntities [["Car", "Motorcycle", "Tank", "StaticWeapon", "Air", "Ship"], _fragRange];  
  
{  
    _targets append crew _x;  
    _targets pushBack _x;  
} forEach _vehicles; 
```
Splitting the arrays ends up always being a slower option